### PR TITLE
Event based communication over sockets.

### PR DESCRIPTION
**Pull request by navin22**
## Description
- Using Event based communication over sockets using LengthPrefix communication channel.
- Channel will now notify when the message is received. 
- Socket client and server and use one thread each, instead of using new thread at every read.
- Related tests and added additional tests.
- ToDo : Only main flow is converted to the socket client and server. Other flows like DataCollection is not on this.

### CODE REVIEW

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 1]
```diff
@@ -6,460 +6,542 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
         /// Initializes a new instance of the <see cref="TestRequestSender"/> class.
         /// </summary>
-        /// <param name="protocolConfig">Protocol related information</param>
+        /// <param name="protocolConfig">Protocol configuration.</param>
         /// <param name="connectionInfo">Transport layer to set up connection</param>
         public TestRequestSender(ProtocolConfig protocolConfig, TestHostConnectionInfo connectionInfo)
-            : this(new SocketCommunicationManager(), connectionInfo, JsonDataSerializer.Instance, protocolConfig)
+            : this(connectionInfo, JsonDataSerializer.Instance, protocolConfig, CLIENTPROCESSEXITWAIT)
         {
+            if (this.connectionInfo.Role == ConnectionRole.Client)
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: The connectioninfo here is that of the RuntimeProvider, which tells how the testhost will act, so we should reverse the role the of vstest.console right?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs [hunk 2]
```diff
@@ -197,22 +197,17 @@ public void Close()
                 return true;
             }
 
-            // In case of DataCollection we only start dc.exe on initialize, & close once the TestRun is complete,
-            // So instead of resuing ProxyExecutionManager, we will close it here, & create new PEMWDC
-            // In Case of Abort, clean old one and create new proxyExecutionManager in place of old one.
-            if (!this.SharedHosts || testRunCompleteArgs.IsAborted || (proxyExecutionManager is ProxyExecutionManagerWithDataCollection))
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs]: Why did we remove these checks? were they bogus?

**Review by mayankbansal018 — COMMENTED** [review]
not taken a look at tests yet, will do it later

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs [hunk 3]
```diff
@@ -41,16 +41,26 @@ protected SocketClient(Func<Stream, ICommunicationChannel> channelFactory)
...
-        public event EventHandler<ConnectedEventArgs> ServerConnected;
+        public event EventHandler<ConnectedEventArgs> Connected;
 
         /// <inheritdoc />
-        public event EventHandler<DisconnectedEventArgs> ServerDisconnected;
+        public event EventHandler<DisconnectedEventArgs> Disconnected;
 
         /// <inheritdoc />
-        public void Start(string connectionInfo)
+        // public void Start(string connectionInfo)
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: Done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs [hunk 4]
```diff
@@ -65,22 +75,28 @@ public void Stop()
 
         private void OnServerConnected(Task connectAsyncTask)
         {
-            if (connectAsyncTask.IsFaulted)
+            // if (connectAsyncTask.IsFaulted)
+            // {
+            //    throw connectAsyncTask.Exception;
+            // }
+            if (this.Connected != null)
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: It's an event so it can be null. But in our case it will not be null.

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs (hunk 1 above)
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: Yeah, it happens in another constructor. Ln 92-94.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs (hunk 2 above)
Comment (navin22) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs]: Earlier they were needed only for those scenarios, and in other scenarios they were reusing the execution manager. But since the test host dies it's not good to have that around. and was observing an exception, which tries to send message over a channel that's already closed.

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs (hunk 1 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: Please put a comment to move this to Communication factory, which returns ICommEndPoint based on Role, & ConnectionType (Socket, NamedPipe etc..) Once another communication mechanism is implemented

**Comment by navin22** [comment]
@dotnet-bot test Windows_NT / Release Build

**Comment by navin22** [comment]
@dotnet-bot test

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs [hunk 5]
```diff
@@ -65,22 +68,24 @@ public void Stop()
...
-                throw connectAsyncTask.Exception;
-            }
-
-            this.channel = this.channelFactory(this.tcpClient.GetStream());
-            if (this.ServerConnected != null)
+            if (this.Connected != null)
             {
-                this.ServerConnected.SafeInvoke(this, new ConnectedEventArgs(this.channel), "SocketClient: ServerConnected");
+                if (connectAsyncTask.IsFaulted)
+                {
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: Add EqtTrace on failure and success scenarios.

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs [hunk 6]
```diff
@@ -19,291 +19,195 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
 
-        /// <summary>
-        /// Listens to the commands from server
-        /// </summary>
-        /// <param name="testHostManagerFactory">the test host manager.</param>
-        public void ProcessRequests(ITestHostManagerFactory testHostManagerFactory)
-        {
-            bool isSessionEnd = false;
+            if (this.connectionInfo.Role == ConnectionRole.Host)
+            {
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: Logging.

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs [hunk 7]
```diff
@@ -19,291 +19,195 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
 
-            do
+        /// <inheritdoc />
+        public virtual void InitializeCommunication()
+        {
+            this.communicationEndPoint.Connected += (sender, connectedArgs) =>
             {
-                var message = this.communicationManager.ReceiveMessage();
-                if (EqtTrace.IsInfoEnabled)
+                if (!connectedArgs.Connected)
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: Logging.

File: test/Microsoft.TestPlatform.Common.UnitTests/Microsoft.TestPlatform.Common.UnitTests.csproj [hunk 8]
```diff
@@ -10,31 +10,20 @@
     <TargetFrameworks>netcoreapp1.0;net451</TargetFrameworks>
     <AssemblyName>Microsoft.TestPlatform.Common.UnitTests</AssemblyName>
   </PropertyGroup>
-  <ItemGroup>
+  <!--<ItemGroup>
```
Comment (smadala) [inline test/Microsoft.TestPlatform.Common.UnitTests/Microsoft.TestPlatform.Common.UnitTests.csproj]: Nit: Remove comment.

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests.csproj [hunk 9]
```diff
@@ -7,12 +7,12 @@
   <Import Project="$(TestPlatformRoot)scripts/build/TestPlatform.Settings.targets" />
   <PropertyGroup>
     <OutputType Condition=" '$(TargetFramework)' == 'netcoreapp1.0' ">Exe</OutputType>
-    <TargetFrameworks>netcoreapp1.0;net451</TargetFrameworks>
+    <TargetFrameworks>net451;netcoreapp1.0</TargetFrameworks>
```
Comment (smadala) [inline test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests.csproj]: Revert the change.

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests.csproj [hunk 10]
```diff
@@ -7,12 +7,12 @@
...
     <AssemblyName>Microsoft.TestPlatform.CommunicationUtilities.PlatformTests</AssemblyName>
     <EnableCodeAnalysis>true</EnableCodeAnalysis>
   </PropertyGroup>
   <ItemGroup>
-    <ProjectReference Include="..\..\src\Microsoft.TestPlatform.CommunicationUtilities\Microsoft.TestPlatform.CommunicationUtilities.csproj" />
+    <!--<ProjectReference Include="..\..\src\Microsoft.TestPlatform.CommunicationUtilities\Microsoft.TestPlatform.CommunicationUtilities.csproj" />
```
Comment (smadala) [inline test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests.csproj]: Remove all the commented code.

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketTestsBase.cs [hunk 11]
```diff
@@ -70,5 +62,15 @@ protected static void WriteData(TcpClient client)
         }
 
         protected abstract ICommunicationChannel SetupChannel(out ConnectedEventArgs connectedEventArgs);
+
+        protected IPEndPoint GetIpEndPoint(string value)
+        {
+            if (Uri.TryCreate(string.Concat("tcp://", value), UriKind.Absolute, out Uri uri))
```
Comment (smadala) [inline test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketTestsBase.cs]: DRY.

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Microsoft.TestPlatform.CommunicationUtilities.UnitTests.csproj [hunk 12]
```diff
@@ -7,12 +7,12 @@
   <Import Project="$(TestPlatformRoot)scripts/build/TestPlatform.Settings.targets" />
   <PropertyGroup>
     <OutputType Condition=" '$(TargetFramework)' == 'netcoreapp1.0' ">Exe</OutputType>
-    <TargetFrameworks>netcoreapp1.0;net451</TargetFrameworks>
+    <TargetFrameworks>net451;netcoreapp1.0</TargetFrameworks>
```
Comment (smadala) [inline test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/Microsoft.TestPlatform.CommunicationUtilities.UnitTests.csproj]: Revert this change(And other places too).

**Review by smadala — CHANGES_REQUESTED** [review]

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketServer.cs [hunk 13]
```diff
@@ -17,7 +17,7 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
     /// <summary>
     /// Communication server implementation over sockets.
     /// </summary>
-    public class SocketServer : ICommunicationServer
+    public class SocketServer : ICommunicationEndPoint
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketServer.cs]: Done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs (hunk 5 above)
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: Done

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketServer.cs [hunk 14]
```diff
@@ -90,10 +88,10 @@ private void OnClientConnected(TcpClient client)
             this.tcpClient = client;
             this.tcpClient.Client.NoDelay = true;
 
-            if (this.ClientConnected != null)
+            if (this.Connected != null)
             {
                 this.channel = this.channelFactory(this.tcpClient.GetStream());
-                this.ClientConnected.SafeInvoke(this, new ConnectedEventArgs(this.channel), "SocketServer: ClientConnected");
+                this.Connected.SafeInvoke(this, new ConnectedEventArgs(this.channel), "SocketServer: ClientConnected");
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketServer.cs]: Done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 15]
```diff
@@ -6,460 +6,543 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
     /// <summary>
-    /// Utility class that facilitates the IPC comunication. Acts as server.
+    /// Test request sender implementation.
     /// </summary>
-    public sealed class TestRequestSender : ITestRequestSender
+    public class TestRequestSender : ITestRequestSender
     {
-        private ICommunicationManager communicationManager;
+        // Time to wait for test host exit (in seconds)
+        private const int CLIENTPROCESSEXITWAIT = 10 * 1000;
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: Done

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 16]
```diff
@@ -6,460 +6,543 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
-            return this.transport.WaitForConnection(clientConnectionTimeout);
-        }
+            this.communicationEndpoint.Connected += (sender, args) =>
+                {
+                    this.channel = args.Channel;
+                    this.connected.Set();
+                };
+            this.communicationEndpoint.Disconnected += (sender, args) =>
+                {
+                    // If there's an disconnected event handler, call it
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: NA

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs [hunk 17]
```diff
@@ -19,291 +19,195 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
-        private ITransport transport;
+        private TestHostConnectionInfo connectionInfo;
 
-        private IDataSerializer dataSerializer;
+        private int highestSupportedVersion = 2;
+        private JobQueue<Action> jobQueue;
+        private ICommunicationChannel channel;
 
-        private Action<Message> onAckMessageRecieved;
+        private ManualResetEventSlim requestSenderConnected;
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: Readability wise requestSenderConnected makes sense. As WaitForTestRequestSenderConnected is using it.

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 18]
```diff
@@ -6,460 +6,543 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
         }
 
-        /// <summary>
-        /// Send the cancel message to test host
-        /// </summary>
+        /// <inheritdoc />
         public void SendTestRunCancel()
         {
-            this.communicationManager.SendMessage(MessageType.CancelTestRun);
+            this.channel?.Send(this.dataSerializer.SerializeMessage(MessageType.CancelTestRun));
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: If this is called before the initialization has happened. I think it's covered wherever it's necessary.

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 19]
```diff
@@ -6,460 +6,543 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
-                    {
-                        var testCases = this.dataSerializer.DeserializePayload<IEnumerable<TestCase>>(message);
-                        discoveryEventsHandler.HandleDiscoveredTests(testCases);
-                    }
-                    else if (string.Equals(MessageType.DiscoveryComplete, message.MessageType))
+                    this.OnDiscoveryAbort(discoveryEventsHandler, disconnectedEventArgs.Error, true);
+                };
+            this.onMessageReceived = (sender, args) =>
+                {
+                    try
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: Done

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs [hunk 20]
```diff
@@ -41,16 +41,19 @@ protected SocketClient(Func<Stream, ICommunicationChannel> channelFactory)
...
-        public void Start(string connectionInfo)
+        public string Start(string endpoint)
         {
-            this.tcpClient.ConnectAsync(IPAddress.Loopback, int.Parse(connectionInfo))
-                .ContinueWith(this.OnServerConnected);
+            var ipEndPoint = endpoint.GetIPEndPoint();
+
+            // Don't start if the endpoint port is zero
+            this.tcpClient.ConnectAsync(ipEndPoint.Address, ipEndPoint.Port).ContinueWith(this.OnServerConnected);
+            return endpoint;
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: just wanted to return the endpoint it's connected.

File: src/Microsoft.TestPlatform.CommunicationUtilities/LengthPrefixCommunicationChannel.cs [hunk 21]
```diff
@@ -57,10 +57,7 @@ public Task NotifyDataAvailable()
             // connection is closed.
             var data = this.reader.ReadString();
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/LengthPrefixCommunicationChannel.cs]: It's caught in MessageLoopAsync method.

File: src/Microsoft.TestPlatform.Common/Interfaces/Communication/ConnectedEventArgs.cs [hunk 22]
```diff
@@ -24,11 +24,28 @@ public ConnectedEventArgs()
         public ConnectedEventArgs(ICommunicationChannel channel)
         {
             this.Channel = channel;
+            this.Connected = true;
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.Common/Interfaces/Communication/ConnectedEventArgs.cs]: Moved the files.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Interfaces/ICommunicationEndpoint.cs [hunk 23]
```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.Interfaces
+{
+    using System;
+
+    public interface ICommunicationEndPoint
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/Interfaces/ICommunicationEndpoint.cs]: Sent it.

[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/Interfaces/ICommunicationEndpoint.cs]: Can be put it in vstest-docs, once refined.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs [hunk 24]
```diff
@@ -41,16 +41,19 @@ protected SocketClient(Func<Stream, ICommunicationChannel> channelFactory)
...
-        public event EventHandler<ConnectedEventArgs> ServerConnected;
+        public event EventHandler<ConnectedEventArgs> Connected;
 
         /// <inheritdoc />
-        public event EventHandler<DisconnectedEventArgs> ServerDisconnected;
+        public event EventHandler<DisconnectedEventArgs> Disconnected;
 
         /// <inheritdoc />
-        public void Start(string connectionInfo)
+        public string Start(string endpoint)
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: Done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: Done.

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 25]
```diff
@@ -6,460 +6,543 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
+                if (!protocolNegotiated.WaitOne(60 * 1000))
+                {
+                    throw new TestPlatformException(string.Format(CultureInfo.CurrentUICulture, CommonResources.VersionCheckTimedout));
+                }
             }
-            else
+            finally
             {
-                throw new TestPlatformException(string.Format(CultureInfo.CurrentUICulture, CommonResources.UnexpectedMessage, MessageType.VersionCheck, message.MessageType));
+                this.channel.MessageReceived -= this.onMessageReceived;
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: Reading from binary reader only when there's a corresponding event handler.

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs [hunk 26]
```diff
@@ -312,13 +216,151 @@ private ITestCaseEventsHandler GetTestCaseEventsHandler(string runSettings)
...
             }
 
             return testCaseEventsHandler;
         }
+
+        public void OnMessageReceived(object sender, MessageReceivedEventArgs messageReceivedArgs)
+        {
+            var message = this.dataSerializer.DeserializeMessage(messageReceivedArgs.Data);
+
+            switch (message.MessageType)
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: Added it back.

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs [hunk 27]
```diff
@@ -1,3 +1,4 @@
+
```
[the remark this answers is not in the document]
Comment (navin22) [reply src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: Done

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs [hunk 28]
```diff
@@ -19,306 +19,367 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
-                var message = this.communicationManager.ReceiveMessage();
-                if (EqtTrace.IsInfoEnabled)
+                if (!connectedArgs.Connected)
                 {
-                    EqtTrace.Info("TestRequestHandler.ProcessRequests: received message: {0}", message);
+                    requestSenderConnected.Set();
+                    throw connectedArgs.Fault;
                 }
+                this.channel = connectedArgs.Channel;
+                this.channel.MessageReceived += this.OnMessageReceived;
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: As soon as we set handler to receive messages, we can get a Test Discovery/Execution request before ProcessRequest which supply testHostManagerFactory.

So move attaching to event handler in ProcessRequest

Again this is a corner scenario, but can happen

**Review by mayankbansal018 — CHANGES_REQUESTED** [review]

**Comment by navin22** [comment]
@dotnet-bot test Windows_NT / Debug Build

**Comment by navin22** [comment]
@dotnet-bot test Windows_NT / Release Build

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs (hunk 28 above)
Comment (navin22) [reply src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: You are right, there some unit test failing randomly because of this.I am handling this in the latest one. Please take a look.

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 29]
```diff
@@ -6,460 +6,550 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
-                    LastDiscoveredTests = null,
-                    TotalTests = -1
-                };
-                rawMessage = this.dataSerializer.SerializePayload(MessageType.DiscoveryComplete, payload);
-                eventHandler.HandleRawMessage(rawMessage);
+            // notify of a test run complete and bail out.
+            testRunEventsHandler.HandleTestRunComplete(completeArgs, null, null, null);
+        }
+
+        private void OnDiscoveryAbort(ITestDiscoveryEventsHandler2 eventHandler, Exception exception, bool getClientError)
```
Comment (cltshivash) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: >OnDiscoveryAbort [](start = 21, length = 16)

exception handling has been removed.. any  impact due to this ?

File: src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs [hunk 30]
```diff
@@ -19,306 +19,375 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
```
Comment (cltshivash) [inline src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs]: >MessageReceived [](start = 29, length = 15)

can the processing of a second message begin before the completion of an earlier one..? what is the impact..?

**Review by mayankbansal018 — APPROVED** [review]

**Comment by mayankbansal018** [comment]
@dotnet-bot test Windows_NT / Release Build

**Comment by mayankbansal018** [comment]
@dotnet-bot test Windows_NT / Release Build

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs [hunk 31]
```diff
@@ -6,460 +6,550 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
...
-            if (message.MessageType == MessageType.VersionCheck)
-            {
-                this.protocolVersion = this.dataSerializer.DeserializePayload<int>(message);
+                    EqtTrace.Info(@"TestRequestSender: VersionCheck Succeeded, NegotiatedVersion = {0}", this.protocolVersion);
+                }
+                else if (message.MessageType == MessageType.ProtocolError)
+                {
+                    throw new TestPlatformException(string.Format(CultureInfo.CurrentUICulture, CommonResources.VersionCheckFailed));
+                }
+                else
```
[the remark this answers is not in the document]
Comment (mayankbansal018) [reply src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs]: added handler for message type MessageType.TestMessage

**Review by smadala — APPROVED** [review]
